sections/tutorials/asyncio/README.md

```diff
@@ -0,0 +1,211 @@
+
+- Code actually runs in parallel.
+
+  _While this might be interesting to "speed up" your processing, it is also very easy to hit problems if you access the same resources from different threads. It is especially an issue in the BGE, since there are specific times at which the BGE is "ready" to execute logic code, but if your run functions in separate threads, you might call APIs outside of these "windows" and the BGE might crash or raise an exception in your code running in the thread._
+
+Threads are not inherently bad, but they must be used with a lot of care, especially in the BGE.
+
+So how can an event-loop help you?
+
+First of all, it can run on just one thread, but allow you to start multiple operations at the same time, without blocking waiting for an answer.
+
+Let's take a look at how a synchronous code may look like (script for a Python Controller in "module" mode):
+
+```py
+# Echo client program
+import socket
+
+HOST = 'daring.cwi.nl'    # The remote host
+PORT = 50007              # The same port as used by the server
+
+def pycontroller_on_event(controller):
+    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as client_socket:
+        client_socket.connect((HOST, PORT))
+        client_socket.sendall(b'Hello, world')
+        data = client_socket.recv(1024)
+
+    print('Received', repr(data))
+```
+
+This code looks fairly simple, it basically creates a socket, sends a message (`.sendall`), and finally waits for a response (`.recv`).
+
+The issue here is that the whole program is blocked on the receive statement. Because of the way it is written, you don't want your program to continue if you did not receive anything, that wouldn't make any sense. So it blocks.
+
+That's the price of writing code for only one thread, if it blocks, you'll have to wait.
+
+In the BGE this would translate into your game freezing, because it will wait for your script to finish, which itself waits for a response.
+
+To solve this issue, you have three solutions:
+
+1. Make the socket non-blocking.
+
+   This will make the `.recv` call return and raise an error if there was nothing to read.
+   You need to write your code to branch off and wait for the next execution to try and read again, until something can be read.
+   It works, but throwing errors on each logic frame and re-executing the same code over and over until it works is a bit annoying to reason about.
+
+   ```py
+   # Echo client program (non-blocking)
+   import socket
+
+   HOST = 'daring.cwi.nl'    # The remote host
+   PORT = 50007              # The same port as used by the server
+
+   def pycontroller_on_event(controller):
+       owner = controller.owner
+
+       client_socket = owner.get('socket', None)
+
+       # Test if someone already expects a response somehow
+       if client_socket is None:
+            client_socket = owner['socket'] = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+            s.settimeout(0)

```

Comment:
Did you define s somewhere?  I apologize if I missed it.

Reply:
Nice catch, I did not. I meant `client_socket`, will fix.